### DISCUSSION TITLE
Type argument inference failing with unambiguous method group

### DISCUSSION THREAD

**Discussion by MorenoGentili** [dotnet/csharplang#5963] [Q&A] [answered]
C# 10 [introduced natural types for lambdas and method groups](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-expressions#natural-type-of-a-lambda-expression). However, compilation error CS0411 still occurs when an unambiguous method group is passed as an argument to a generic method. Please consider this case, as discussed in https://github.com/dotnet/csharplang/discussions/129

```csharp
static bool IsEven(int x) => x % 2 == 0;

static void Test<T>(Func<T, bool> predicate)
{
}

// type inference succeeds here
var isEven = IsEven;

// But fails here, why?
Test(IsEven);
```
If the `var` assignment works, then the method invocation should as well.

_Originally posted by @MorenoGentili in https://github.com/dotnet/csharplang/discussions/129#discussioncomment-2455671_

**Comment by RikkiGibson** [comment]
Changing `Test(IsEven)` to `Test(isEven)` makes the error disappear. [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQAcoKgWzczAOwBoYQBLAG2IBMQBqAHwAEAmARgFgAoH59gGwACYAHtRlIQElkAUQBuEQgApyhGEIAeASiEBeAHxahAUiGt9eoQAYA3Dz6ChzACxCAKihgAedweXMAKy+xCLilEboSDTkAMawENo8AN48AL4O3AD0WUIwAJ7oEEJqkEiEscXIcLGVEDTIQgAWEEg88lglcoqE+tLdSva82bkAQnAaYFBUjS1IoQDuTfkA/DyeyDCqA4TatkA=). This makes me suspect that this is a compiler bug. It might be best to file an issue in roslyn.

Part of the problem might be that this was actually implemented as a "special" conversion which occurs when there's no target type or the target type is `object` or `System.Delegate`. But it still seems like the result type of this "conversion" should be able to contribute to type inference. cc @cston.

**Reply by 333fred** [reply]
I doubt this was unintentional. @cston, was this one of the backcompat things we did?

**Comment by cston** [comment, marked as answer]
The local `isEven` has a _delegate_type_, and type inference for `Test(isEven)` will infer from the type arguments of the local type `Func<int, bool>` to the type parameters of the target type `Func<T, bool>`: see [spec](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#116310-lower-bound-inferences).

However, the _method group_ `IsEven` has a [_function_type_](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/lambda-improvements.md#natural-function-type) rather than a _delegate_type_, and there is no type inference rule for inferring from the _parameter types_ of a _function_type_ to the type parameters of the target type.

(By contrast, type inference of the corresponding _lambda expression_ `Test((int x) => x % 2 == 0);` succeeds not because of _function_type_ of the lambda but because type inference has a specific rule for anonymous functions: see [spec](https://github.com/dotnet/csharpstandard/blob/draft-v6/standard/expressions.md#11638-explcit-parameter-type-inferences).)

**Reply by MorenoGentili** [reply]
Thanks @cston for your detailed explanation.
> because type inference has a specific rule for anonymous function

Would you consider adding a specific rule for method groups as well?

**Reply by jnm2** [reply]
(Which is what https://github.com/dotnet/csharplang/discussions/129 has been asking for)

**Reply by RikkiGibson** [reply]
It wasn't clear to me if there was a specific reason this scenario wasn't supported as part of the lambda improvements in C# 10. For example, is there a compat concern or technical hurdle that makes it impractical?
